diff --git a/bl31/bl31_main.c b/bl31/bl31_main.c
index dea50d422..46e11b2bf 100644
--- a/bl31/bl31_main.c
+++ b/bl31/bl31_main.c
@@ -134,6 +134,7 @@ void bl31_main(void)
 
 	NOTICE("BL31: %s\n", build_version_string);
 	NOTICE("BL31: %s\n", build_message);
+	NOTICE("OMNIVISOR PATCH\n");
 
 #if FEATURE_DETECTION
 	/* Detect if features enabled during compilation are supported by PE. */
diff --git a/include/services/std_svc.h b/include/services/std_svc.h
index b0614fb42..5cb661417 100644
--- a/include/services/std_svc.h
+++ b/include/services/std_svc.h
@@ -14,6 +14,10 @@
 /*					0x8400ff02 is reserved */
 #define ARM_STD_SVC_VERSION		0x8400ff03
 
+/* SVCs to read EL3 mapped registers on ZCU102/KRIA */
+#define ARM_STD_SVC_READ_REG		0x8400ff04
+#define ARM_STD_SVC_WRITE_REG		0x8400ff05
+
 /* ARM Standard Service Calls version numbers */
 #define STD_SVC_VERSION_MAJOR		0x0
 #define STD_SVC_VERSION_MINOR		0x1
diff --git a/plat/xilinx/zynqmp/include/zynqmp_def.h b/plat/xilinx/zynqmp/include/zynqmp_def.h
index cd3bbbc64..9f5bbae5e 100644
--- a/plat/xilinx/zynqmp/include/zynqmp_def.h
+++ b/plat/xilinx/zynqmp/include/zynqmp_def.h
@@ -42,14 +42,20 @@
  * ZYNQMP memory map related constants
  ******************************************************************************/
 /* Aggregate of all devices in the first GB */
-#define DEVICE0_BASE		U(0xFF000000)
-#define DEVICE0_SIZE		U(0x00E00000)
+/* XMPU support ZCU: extend DEVICE0 do cover XMPU registers 
+ * FIXME: does this generate problems?
+ */
+#define DEVICE0_BASE		U(0xFD000000)	//was 0xFF000000
+#define DEVICE0_SIZE		U(0x02E00000)	//was 0x00E00000
 #define DEVICE1_BASE		U(0xF9000000)
 #define DEVICE1_SIZE		U(0x00800000)
 
 /* For cpu reset APU space here too 0xFE5F1000 CRF_APB*/
 #define CRF_APB_BASE		U(0xFD1A0000)
-#define CRF_APB_SIZE		U(0x00600000)
+/* QoS support ZCU 102: extend do cover FPD_GPV and LPD_GPV
+ * FIXME: does this generate problems with the IOMMU?
+ */
+#define CRF_APB_SIZE		U(0x02000000)
 #define CRF_APB_CLK_BASE	U(0xFD1A0020)
 
 /* CRF registers and bitfields */
diff --git a/services/std_svc/std_svc_setup.c b/services/std_svc/std_svc_setup.c
index deca1c044..874cf3701 100644
--- a/services/std_svc/std_svc_setup.c
+++ b/services/std_svc/std_svc_setup.c
@@ -13,6 +13,7 @@
 #include <lib/pmf/pmf.h>
 #include <lib/psci/psci.h>
 #include <lib/runtime_instr.h>
+#include <lib/mmio.h>
 #include <services/drtm_svc.h>
 #include <services/errata_abi_svc.h>
 #include <services/pci_svc.h>
@@ -233,6 +234,16 @@ static uintptr_t std_svc_smc_handler(uint32_t smc_fid,
 		/* Return the version of current implementation */
 		SMC_RET2(handle, STD_SVC_VERSION_MAJOR, STD_SVC_VERSION_MINOR);
 
+	case ARM_STD_SVC_READ_REG:
+		//NOTICE(" X1: 0x%lx\n", x1);
+		SMC_RET1(handle, mmio_read_32(x1));
+
+	case ARM_STD_SVC_WRITE_REG:
+		//NOTICE("BEFORE X1: 0x%lx X2: 0x%lx\n", x1, x2);
+		mmio_write_32(x1, x2);
+		//NOTICE("AFTER  X1: 0x%lx X2: 0x%lx\n", x1, x2);
+		SMC_RET1(handle, mmio_read_32(x1));
+
 	default:
 		VERBOSE("Unimplemented Standard Service Call: 0x%x \n", smc_fid);
 		SMC_RET1(handle, SMC_UNK);
