diff --git a/drivers/remoteproc/remoteproc_core.c b/drivers/remoteproc/remoteproc_core.c
index 001d75244..9b0794615 100644
--- a/drivers/remoteproc/remoteproc_core.c
+++ b/drivers/remoteproc/remoteproc_core.c
@@ -43,6 +43,7 @@
 #include "remoteproc_internal.h"
 
 #define HIGH_BITS_MASK 0xFFFFFFFF00000000ULL
+#define CONFIG_OMNV_RPROC 1
 
 static DEFINE_MUTEX(rproc_list_mutex);
 static LIST_HEAD(rproc_list);
@@ -1297,6 +1298,10 @@ static int rproc_start(struct rproc *rproc, const struct firmware *fw)
 		goto reset_table_ptr;
 	}
 
+#ifdef CONFIG_OMNV_RPROC
+	return 0;
+#endif
+
 	/* power up the remote processor */
 	ret = rproc->ops->start(rproc);
 	if (ret) {
